tower/src/retry/backoff.rs

```diff
@@ -0,0 +1,228 @@
+//! This module contains generic backoff utlities to be used with the retry
+//! layer.
+//!
+//! The [`Backoff`] trait is a generic way to represent backoffs that can use
+//! any timer type.
+//!
+//! [`ExponentialBackoff`] implements [`Backoff`] and provides a batteries
+//! included exponential backoff and jitter strategy.
+
+use rand::Rng;
+use rand::{rngs::SmallRng, thread_rng, SeedableRng};
+use std::fmt::Display;
+use std::future::Future;
+use std::time::Duration;
+use tokio::time;
+
+/// A backoff trait where a single mutable refernces represents a single
+/// backoff session. Implementors must also implement [`Clone`] which will
+/// reset the backoff back to the default state for the next session.
+pub trait Backoff: Clone {
+    /// The future associated with each backoff. This usually will be some sort
+    /// of timer.
+    type Future: Future<Output = ()>;
+
+    /// Initiate the next backoff in the sequence.
+    fn next_backoff(&mut self) -> Self::Future;
+}
+
+/// A jittered exponential backoff strategy.
+#[derive(Debug)]
+pub struct ExponentialBackoff<R = SmallRng> {
+    /// The minimum amount of time to wait before resuming an operation.
+    min: time::Duration,
+
+    /// The maximum amount of time to wait before resuming an operation.
+    max: time::Duration,
+
+    /// The ratio of the base timeout that may be randomly added to a backoff.
+    ///
+    /// Must be greater than or equal to 0.0.
+    jitter: f64,
+
+    rng: R,
+    iterations: u32,
+}
+
+impl<R> ExponentialBackoff<R>
+where
+    R: Rng,
```

Comment:
I'd suggest people not depend on `rand::Rng`. It's fairly straightforward to make a simple internal RNG by making use of [`RandomState`](https://doc.rust-lang.org/std/collections/hash_map/struct.RandomState.html) and a counter.

Reply:
So one thing is how can we have deterministic testing of retries if `RandomState` is buried in this type?